includes/Shopping/Shopify_Query.php

```diff
@@ -214,55 +225,103 @@ protected function fetch_remote_products( string $search_term, string $orderby,
 		 * @param int $time Time to live (in seconds). Default is 5 minutes.
 		 */
 		$cache_ttl = apply_filters( 'web_stories_shopify_data_cache_ttl', 5 * MINUTE_IN_SECONDS );
-		$cache_key = 'web_stories_shopify_data_' . md5( $search_term . '-' . $orderby . '-' . $order );
+		$cache_key = 'web_stories_shopify_data_' . md5( $search_term . '-' . $page . '-' . $per_page . '-' . $orderby . '-' . $order );
 
 		$data = get_transient( $cache_key );
 
 		if ( \is_string( $data ) && ! empty( $data ) ) {
 			return (array) json_decode( $data, true );
 		}
 
-		$query = $this->get_products_query( $search_term, $orderby, $order );
+		$after = '';
+		if ( $page > 1 ) {
+			for ( $i = 1; $i < $page; $i ++ ) {
+				$query = $this->get_products_query( $search_term, $after, $per_page, $orderby, $order );
+				$body  = $this->execute_query( $query );
+				if ( is_wp_error( $body ) ) {
+					return $body;
+				}
+				$validate = $this->validate_request( $body );
+				if ( is_wp_error( $validate ) ) {
+					return $validate;
+				}
+				/**
+				 * Shopify GraphQL API response.
+				 *
+				 * @var ShopifyGraphQLResponse $result
+				 */
+				$result        = json_decode( $body, true );
+				$has_next_page = $result['data']['products']['pageInfo']['hasNextPage'];
+				if ( ! $has_next_page ) {
+					return new WP_Error( 'rest_no_page', __( 'Error fetching products', 'web-stories' ), [ 'status' => 404 ] );
+				}
+				$after = (string) $result['data']['products']['pageInfo']['endCursor'];
+			}
+		}
```

Comment:
This causes quite a lot of requests to the Shopify API.

In the editor, when paginating through the products responses, for every page N it would make N API requests to Shopify. Example: when you reach page 10 in the editor, it will call `/web-stories/v1/products?page=10&per_page=20`, which then leads to `execute_query()` being called 10 times here.

1. Page 1: 1 request
2. Page 2: 2 requests
3. Page 3: 3 requests
4. and so on

When reaching page 10, there will have been 55 API requests in a very short time. This way users will very quickly run into [API rate limits](https://shopify.dev/api/usage/rate-limits#time-based-limits).

What would it take for our API to support cursor-based pagination as well?

Then, when paginating in the editor we would just call `/web-stories/v1/products?per_page=20&after=abc`, `/web-stories/v1/products?per_page=20&after=def`, etc. This will result in way fewer API requests.

Reply:
I think that supporting cursor based pagination breaks the genericness we have at the moment. I think we need talk about this change, as it is fundamental.

Reply:
Yeah it would definitely mean updating the current genericness to account for both types of pagination.

I'd imagine the editor would make calls like `/web-stories/v1/products?per_page=20&page=2&after=abc` and we'd pass both `$page` and `$after` to the `Product_Query`.

WooCommerce would just take `$page` and ignore `$after`, and Shopify would ignore `$page` instead and use `$after`.

That's why I am curious to see how much work that would be.

----

Regardless, it looks like `get_search()` needs to return not just an array of `Product`s, but also pagination information (`hasNextPage` and `endCursor` for Shopify`, `max_num_pages`, `total` for WooCommerce)

Then we need to add this information as REST API response headers (e.g. `X-WP-Total`, `X-WP-TotalPages`, `X-WP-EndCursor`)

WooCommerce supports this with the `'paginate' => true` flag, see https://github.com/woocommerce/woocommerce/wiki/wc_get_products-and-WC_Product_Query

```php
// Get products with extra info about the results.
$args = array(
    'paginate' => true,
);
$results = wc_get_products( $args );
echo $results->total . ' products found\n';
echo 'Page 1 of ' . $results->max_num_pages . '\n';
echo 'First product id is: ' . $results->products[0]->get_id() . '\n';
```

Reply:
I have refactored to move the caching. I think we don't need to cursor. The current implemention, means when you hit, page 1 and the page 2, it will reuse page 1, meaning no more requests.

Reply:
Okay, I'll take a look at it.

The next problem is that in the editor we still need to know if there are more pages to fetch. Which means we still need to send some API response headers so that we can then implement pagination in the editor.